core/slim_sim.cpp

```diff
@@ -6284,8 +6229,25 @@ void SLiMSim::WriteTreeSequenceMetadata(tsk_table_collection_t *p_tables)
     // leaving other keys that might already be there.
     // But that's being a headache, so we're skipping it.
     nlohmann::json metadata;
-	metadata["SLiM"]["model_type"] = (ModelType() == SLiMModelType::kModelTypeWF) ? "WF" : "nonWF";
-	metadata["SLiM"]["generation"] = Generation();
+    if (ModelType() == SLiMModelType::kModelTypeWF) {
+        metadata["SLiM"]["model_type"] = "WF";
+        if (GenerationStage() == SLiMGenerationStage::kWFStage1ExecuteEarlyScripts) {
+            metadata["SLiM"]["stage"] = "early";
+        } else {
+            assert(GenerationStage() == SLiMGenerationStage::kWFStage5ExecuteLateScripts);
```

Comment:
I guess this assert is OK-ish.  I don't think we prohibit people from calling treeSeqOutput() from other places – a reproduction() callback, say.  But maybe we should?  Maybe add that to our list of things we should tighten up to make our lives easier?

Reply:
Oh, right, good point. I vote to tighten that up - dealing with that correctly in pyslim will be a pain. The error could even say "if you really want to do this please file a github issue" - but, I can't currently imagine a situation where someone would actually need to do this.